enhancements/aws-pod-identity.md

```diff
@@ -0,0 +1,528 @@
+  external location like an S3 bucket in that case.
+
+## Summary
+
+When deployed on AWS and configured to use [bound service
+tokens](https://github.com/openshift/enhancements/blob/master/enhancements/kube-apiserver/bound-sa-tokens.md),
+Kubernetes pods can assume Roles managed by IAM via a binding created by a
+ServiceAccount annotation.  AWS has a
+[webhook](https://github.com/aws/amazon-eks-pod-identity-webhook/) they maintain
+for EKS that does enables this ability.  This enhancement integrates the same
+functionality into OpenShift.
+
+The `cloud-credential-operator` will be responsible for managing this new
+ability as it serves as an alternative credential _injection_ mechanism versus
+the existing mechanism of injecting an IAM user access ID and key into a
+secret.  The existing role/user _creation_ via a CredentialRequest CR is common
+between the methods.
+
+The new injection mechanism works by creating a `ServiceAccount` annotated with
+a IAM role the pod should assume. When a pod runs with the annotated
+`ServiceAccount`, the webhook injects a projected volume into the pod that
+contains the bound token that can be used to assume the IAM role.  The AWS SDK
+searches for this token automatically.
+
+## Motivation
+
+This allows users to more cleanly express and manage pod identities in IAM when
+running on AWS.  The use of bound tokens to assume roles (vs user id/key) is
+preferred as these tokens can expire and be rotated.
+
+### Goals
+
+1. A pod can assume an IAM role created with a `CredentialRequest` CR using a
+   bounded token
+2. An admin can create a IAM role via `CredentialRequest` CR and inject to a
+   `ServiceAccount` (just as they can create an IAM user and inject to a
+   `Secret` currently)
+4. The token signing CA (`bound-sa-token-signing-certs`) can rotate and should
+   sync between OCP and the OIDC provider in IAM.  Rotation should be seamless.
+   Assumes applications in pods monitor the token for rotation (AWS SDK does
+   this)
+
+### Non-Goals
+
+1. Convert existing IAM user-based components (e.g. ingress, image-registry,
+   machine-api) to use bound tokens.  The migration path will likely be manual
+   if an admin for an existing cluster wishes to move solely to this method.  At
+   some point in the future, we could migrate core components on fresh installs
+   to use this method if we work out the ordering i.e. the webhook needs to be
+   running and the IAM roles need to exist before any of the pods for these
+   components are created and would require change and coordination with the
+   operators that manage those components.
+2. Registration of the OIDC provider in IAM.  EKS leaves this up to the user and
+   it seems the safe thing to do is follow suit.  EKS _does_ create the OIDC
+   endpoint that contains the discovery and keys documents.  Note that there is
+   a hard limit of 1000 OIDC providers in IAM per account.  Leaving the OIDC
```

Comment:
The limit is 100 and not 1000

Reply:
It's the same limit as saml providers in IAM.
https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_iam-limits.html